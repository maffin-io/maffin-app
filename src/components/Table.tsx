import React from 'react';
import {
  ColumnDef,
  TableOptions,
  useReactTable,
  flexRender,
  getCoreRowModel,
  getSortedRowModel,
  getExpandedRowModel,
  ColumnSort,
  ExpandedState,
  getPaginationRowModel,
} from '@tanstack/react-table';
import classNames from 'classnames';
import { FaSortDown, FaSortUp } from 'react-icons/fa';

import Pagination from '@/components/Pagination';

export type TableProps<T extends object> = {
  id: string,
  columns: ColumnDef<T>[],
  data: T[],
  initialSort?: ColumnSort,
  showHeader?: boolean,
  showPagination?: boolean,
  tdClassName?: string,
  isExpanded?: boolean,
} & Partial<TableOptions<T>>;

export default function Table<T extends object = {}>(
  {
    id,
    initialSort,
    showHeader = true,
    showPagination = false,
    tdClassName = 'px-6 py-4',
    isExpanded = false,
    ...props
  }: TableProps<T>,
): JSX.Element {
  const tableConfig: TableOptions<T> = {
    getCoreRowModel: getCoreRowModel(),
    getSortedRowModel: getSortedRowModel(),
    getExpandedRowModel: getExpandedRowModel(),
    enableExpanding: true,
    initialState: {
      ...props.initialState,
      sorting: (initialSort && [initialSort]) || undefined,
      expanded: isExpanded as ExpandedState,
    },
    ...props,
  };

  if (showPagination && !props.manualPagination) {
    tableConfig.getPaginationRowModel = getPaginationRowModel();
  }

  const table = useReactTable(tableConfig);

  return (
    <>
      <div className="relative overflow-hidden rounded-md">
        <table id={id} className="w-full text-sm text-left">
          {
            showHeader
            && (
              <thead className="bg-white dark:bg-dark-700">
                {table.getHeaderGroups().map((headerGroup) => (
                  <tr key={headerGroup.id}>
                    {headerGroup.headers.map(header => (
                      <th
                        key={header.id}
                        scope="col"
                        className="px-6 py-3"
                        onClick={header.column.getToggleSortingHandler()}
                      >
                        <div className="flex">
                          {flexRender(header.column.columnDef.header, header.getContext())}
                          {(
                            header.column.columnDef.header !== ''
                              && header.column.getCanSort()
                              && (
                                <div className="flex items-center cursor-pointer">
                                  <FaSortUp
                                    className={
                                      classNames(
                                        {
                                          'opacity-100': header.column.getIsSorted() === 'asc',
                                          'opacity-50': header.column.getIsSorted() !== 'asc',
                                        },
                                        'absolute',
                                      )
                                    }
                                  />
                                  <FaSortDown
                                    className={
                                      classNames(
                                        {
                                          'opacity-100': header.column.getIsSorted() === 'desc',
                                          'opacity-50': header.column.getIsSorted() !== 'desc',
                                        },
                                        'absolute',
                                      )
                                    }
                                  />
                                </div>
                              )
                          )}
                        </div>
                      </th>
                    ))}
                  </tr>
                ))}
              </thead>
            )
          }

          <tbody>
            {table.getRowModel().rows.map(row => (
              <tr key={row.id} className="border-b border-white/70 dark:border-dark-700">
                {row.getVisibleCells().map(cell => (
                  <td key={cell.id} className={tdClassName}>
                    {flexRender(cell.column.columnDef.cell, cell.getContext())}
                  </td>
                ))}
              </tr>
            ))}
          </tbody>
        </table>
      </div>

      {
        showPagination
        && <Pagination<T> table={table} />
      }
    </>
  );
}
