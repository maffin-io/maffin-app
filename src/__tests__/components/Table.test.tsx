import React from 'react';
import {
  render,
  screen,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { ColumnDef } from '@tanstack/react-table';

import Table from '@/components/Table';

type TestDataType = {
  name: string;
  amount: number;
  extra?: {
    info: string;
  };
};

const columns: ColumnDef<TestDataType>[] = [
  {
    accessorKey: 'name',
    header: 'Name',
  },
  {
    accessorKey: 'amount',
    header: 'Amount',
  },
  {
    header: 'Info',
    accessorFn: (data: TestDataType) => data.extra?.info,
    enableSorting: false,
    cell: ({ getValue }) => (
      <span>
        {getValue<string>()}
      </span>
    ),
  },
];

describe('Table', () => {
  it('renders headers with empty data', () => {
    const { container } = render(
      <Table<TestDataType>
        id="table"
        columns={columns}
        data={[]}
      />,
    );

    screen.getByText('Name');
    screen.getByText('Amount');
    expect(container).toMatchSnapshot();
  });

  it('displays data', () => {
    render(
      <Table<TestDataType>
        id="table"
        columns={columns}
        data={[
          {
            name: 'name1',
            amount: 100,
          },
          {
            name: 'name2',
            amount: 200,
            extra: {
              info: 'info',
            },
          },
        ]}
      />,
    );

    screen.getByText('name1');
    screen.getByText('100');

    screen.getByText('name2');
    screen.getByText('200');
    screen.getByText('info');
  });

  it('can sort', async () => {
    render(
      <Table<TestDataType>
        id="table"
        columns={columns}
        data={[
          {
            name: 'name1',
            amount: 100,
          },
          {
            name: 'name2',
            amount: 200,
            extra: {
              info: 'info',
            },
          },
        ]}
      />,
    );

    let rows = screen.getAllByRole('row');
    expect(rows[1]).toHaveTextContent('name1100');
    expect(rows[2]).toHaveTextContent('name2200info');

    await userEvent.click(screen.getByText('Amount'));

    rows = screen.getAllByRole('row');
    expect(rows[2]).toHaveTextContent('name1100');
    expect(rows[1]).toHaveTextContent('name2200info');
  });

  it('can provide default sort', async () => {
    render(
      <Table<TestDataType>
        id="table"
        columns={columns}
        data={[
          {
            name: 'name2',
            amount: 200,
          },
          {
            name: 'name1',
            amount: 100,
          },
        ]}
        initialSort={{ id: 'name', desc: true }}
      />,
    );

    const rows = screen.getAllByRole('row');
    expect(rows[2]).toHaveTextContent('name1100');
    expect(rows[1]).toHaveTextContent('name2200');
  });
});
