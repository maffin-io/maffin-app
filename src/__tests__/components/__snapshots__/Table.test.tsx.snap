// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table renders headers with empty data 1`] = `
<div>
  <div
    class="relative overflow-hidden"
  >
    <table
      class="w-full text-sm text-left"
      id="table"
    >
      <thead
        class="bg-gunmetal-700 text-slate-300"
      >
        <tr>
          <th
            class="px-6 py-3"
            scope="col"
          >
            <div
              class="flex"
            >
              Name
              <div
                class="flex items-center cursor-pointer"
              >
                <svg
                  class="opacity-50 absolute"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 320 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M279 224H41c-21.4 0-32.1-25.9-17-41L143 64c9.4-9.4 24.6-9.4 33.9 0l119 119c15.2 15.1 4.5 41-16.9 41z"
                  />
                </svg>
                <svg
                  class="opacity-50 absolute"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 320 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41z"
                  />
                </svg>
              </div>
            </div>
          </th>
          <th
            class="px-6 py-3"
            scope="col"
          >
            <div
              class="flex"
            >
              Amount
              <div
                class="flex items-center cursor-pointer"
              >
                <svg
                  class="opacity-50 absolute"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 320 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M279 224H41c-21.4 0-32.1-25.9-17-41L143 64c9.4-9.4 24.6-9.4 33.9 0l119 119c15.2 15.1 4.5 41-16.9 41z"
                  />
                </svg>
                <svg
                  class="opacity-50 absolute"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 320 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41z"
                  />
                </svg>
              </div>
            </div>
          </th>
          <th
            class="px-6 py-3"
            scope="col"
          >
            <div
              class="flex"
            >
              Info
            </div>
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </div>
  <div
    class="flex text-center text-sm items-center px-2 pb-1"
  >
    <label
      class="mr-1"
    >
      Display:
    </label>
    <select>
      <option
        value="5"
      >
        5
      </option>
      <option
        value="10"
      >
        10
      </option>
      <option
        value="50"
      >
        50
      </option>
    </select>
    <span
      class="mx-5"
    >
      Page
       
      <strong>
        1
         
        of
         
        0
      </strong>
       
    </span>
    <span
      class="flex justify-start items-center"
    >
      <label
        class="mr-1"
      >
        Go to page:
      </label>
      <input
        class="w-1/4"
        type="number"
        value="1"
      />
    </span>
    <div
      class="ms-auto"
    >
      <button
        class="bg-gunmetal-700 rounded-full disabled:opacity-50 px-3 py-1"
        disabled=""
        type="button"
      >
        &lt;&lt;
      </button>
       
      <button
        class="bg-gunmetal-700 rounded-full disabled:opacity-50 px-3 py-1"
        disabled=""
        type="button"
      >
        &lt;
      </button>
       
      <button
        class="bg-gunmetal-700 rounded-full disabled:opacity-50 px-3 py-1"
        disabled=""
        type="button"
      >
        &gt;
      </button>
       
      <button
        class="bg-gunmetal-700 rounded-full disabled:opacity-50 px-3 py-1"
        disabled=""
        type="button"
      >
        &gt;&gt;
      </button>
       
    </div>
  </div>
</div>
`;
